docs/source/jobs.rst

```diff
@@ -239,41 +239,137 @@ Use cases for the **job document** include, but are not limited to:
 
 .. code-block:: python
 
     >>> import numpy as np
     >>> job = project.open_job(statepoint)
-    >>> job.data['x'] = np.arange(0, 1, 0.01)
-    >>> job.data['hello'] = 'world'
+    >>> job.data['x'] = np.ones([10, 3, 4])
+
 
 Just like the job *state point* and *document*, individual keys may be accessed either as attributes or through a functional interface, *e.g.*.
-The following examples are all equivalent:
+
+To access data as attributes:
 
 .. code-block:: python
 
-    >>> print(job.data.get('hello'))
-    world
-    >>> print(job.data['hello'])
-    world
-    >>> print(job.data.hello)
-    world
+    >>> with job.data:
+    ...     x = job.data.x[:]
+
+To access data as a key:
+
+.. code-block:: python
+
+    >>> with job.data:
+    ...     x = job.data['x'][:]
+
+Through a functional interface:
+
+.. code-block:: python
+    
+    >>> with job.data:
+    ...     x = job.data.get('x')[:]
 
 .. tip::
 
-     Use the :py:meth:`Job.data.get` method to return ``None`` or another specified default value for missing values. This works exactly like with python's `built-in dictionaries <https://docs.python.org/3/library/stdtypes.html#dict.get>`_.
+     Use the :py:meth:`Job.data.get` method to return ``None`` or another specified default value for missing values. This works exactly like with python's `built-in dictionaries      Use the :py:meth:`Job.data.get` method to return ``None`` or another specified default value for missing values. This works exactly like with python's built-in dictionaries (see :py:meth:`dict.get`). 
+    
+Accessing arrays
+----------------
+
+All values stored in :attr:`job.data` are returned as copies, except for arrays, which are accessed *by reference* and not automatically copied into memory.
+That is important to enable the storage of massive arrays that do not necessarily fit into memory.
+
+For fast and efficient data access, NumPy slicing syntax may be used to access data. 
+Here are a few examples with outputs omitted:
+   
+.. code-block:: python
+    
+    >>> with job.data:
+    ...     job.data.x[0, 0, 0]
+    ...     job.data.x[1:3, 0, :]
+    ...     job.data.x[:, 1, 3]
+    
+To load entire arrays to memory, NumPy slicing syntax may be used:
+ 
+.. code-block:: python
+
+    >>> with job.data:
+    ...     x = job.data.x[:]
+
+You can also create an explicit memory copy of the entire array using the copy-operator ``[()]``:
+
+.. code-block:: python
+
+    >>> with job.data:
+    ...     x = job.data.x[()]
+
+Data organization
+-----------------
+
+The `HDF5`_ format used by :attr:`job.data` allows for hierarchical orgaization of data. 
+Data may be stored in folder-like *groups*:
+
+.. code-block:: python
+    
+    >>> job.data['group/subgroup_1'] = np.ones([10, 3, 2])
+    >>> job.data['group/subgroup_2'] = np.ones([10, 1, 2])
+
+Data may be accessed as attributes, keys, or through a functional interface. 
+The following examples are all equivalent:
+.. code-block:: python
```

Comment:
I think there should be a space above the `code-block` directive. (also in other places)

Reply:
To clarify, an empty line (not a spacebar kind of space) is needed above and below `.. code-block:: python`.